### ISSUE TITLE
Invalid_argument("index out of bounds") in asmcomp/closure.ml when compiling packed module

### ISSUE THREAD

**Issue by copy** [ocaml/ocaml#8769]
Tested on 4.08.0 (no flambda). It compiles fine on 4.07.

```
% cat fortuna.ml 
% cat rng.ml 
module F = Fortuna
% cat nocrypto.mli 
module Rng : sig
  module F : sig end
end
% ocamlc -c -I src -o nocrypto.cmi nocrypto.mli
% ocamlopt -c -for-pack Nocrypto -I src -o fortuna.cmx fortuna.ml
% ocamlopt -c -for-pack Nocrypto -I src -o rng.cmx rng.ml
% ocamlopt -pack -I src rng.cmx -o nocrypto.cmx
Fatal error: exception Invalid_argument("index out of bounds")
Raised by primitive operation at file "asmcomp/closure.ml", line 450, characters 45-50
Called from file "asmcomp/closure.ml", line 1040, characters 28-51
Called from file "asmcomp/closure.ml", line 978, characters 25-53
Called from file "asmcomp/closure.ml", line 1154, characters 27-46
Called from file "asmcomp/closure.ml", line 1056, characters 21-55
Called from file "asmcomp/closure.ml", line 987, characters 31-71
Called from file "asmcomp/closure.ml", line 1154, characters 27-46
Called from file "asmcomp/closure.ml", line 1056, characters 21-55
Called from file "asmcomp/closure.ml", line 987, characters 31-71
Called from file "asmcomp/closure.ml", line 978, characters 25-53
Called from file "asmcomp/closure.ml", line 1438, characters 24-57
Called from file "asmcomp/asmgen.ml", line 225, characters 16-71
Called from file "utils/misc.ml", line 31, characters 8-15
Re-raised at file "utils/misc.ml", line 45, characters 10-56
Called from file "asmcomp/asmgen.ml", line 155, characters 7-231
Called from file "utils/misc.ml", line 31, characters 8-15
Re-raised at file "utils/misc.ml", line 45, characters 10-56
Called from file "asmcomp/asmgen.ml" (inlined), line 261, characters 2-151
Called from file "asmcomp/asmpackager.ml", line 119, characters 6-187
Called from file "utils/misc.ml", line 31, characters 8-15
Re-raised at file "utils/misc.ml", line 45, characters 10-56
Called from file "asmcomp/asmpackager.ml", line 231, characters 2-78
Called from file "utils/misc.ml", line 31, characters 8-15
Re-raised at file "utils/misc.ml", line 45, characters 10-56
Called from file "utils/misc.ml", line 31, characters 8-15
Re-raised at file "utils/misc.ml", line 45, characters 10-56
Called from file "driver/optmain.ml", line 298, characters 6-200
Re-raised at file "parsing/location.ml", line 904, characters 14-25
Called from file "driver/optmain.ml", line 333, characters 6-37
Called from file "driver/optmain.ml", line 337, characters 2-9
```

**Comment by nojb** [comment]
Thanks for the report. The bug was introduced by 111d4e18276dd1fe1605f0b9760d598a201ceb45 @lpw25

**Comment by mshinwell** [comment]
This appears to be a miscompilation of `ocamlopt`, given the offending source line and position is:
```
  | Value_tuple a when n < Array.length a -> a.(n)
                                             ^^^^^
```

**Comment by mshinwell** [comment]
Or maybe `n` is negative, in which case it's not a miscompilation.

**Comment by lpw25** [comment]
Fix in #8770.
